cardano-node: 1.20.0 -> 1.21.1

disassembler opened the pull request:
### Issue Number

None

### Overview

Update for [cardano-node v1.21.1](https://github.com/input-output-hk/cardano-node/releases/tag/1.21.1).

- [x] bumps niv pin for cardano-node.
- [x] Merge stack snapshot yaml for cardano 1.21.1 (PR input-output-hk/cardano-haskell#30).
- [x] Update cardano snapshot version in stack.yaml.
- [x] Fix compile errors.
- [ ] ~Try to import symbols from `Cardano.Api.*` instead of `Shelley.Spec.Ledger.*`, etc, where possible (see diff for new exports).~ (I attempted this and it's rather painful - better to do it later)

### Comments

- [cardano-node diff 1.20.0..1.21.1](https://github.com/input-output-hk/cardano-node/compare/1.20.0...1.21.1)

disassembler committed:
cardano-node: 1.20.0 -> 1.21.1

rvl committed:
stack.yaml: Bump to snapshots/cardano-1.21.1.yaml

https://github.com/input-output-hk/cardano-haskell/pull/30

rvl committed:
Regenerate nix

rvl committed:
Add haddock build workaround

rvl committed:
Build fix for shelley ledger changes

rvl committed:
Handle new MuxError constructors

Anviking reviewed (approved):
lgtm

hasufell reviewed (dismissed):
Cabal config files need to be regenerated, see

https://github.com/input-output-hk/cardano-wallet/wiki/Building#syncing-stack-and-cabal-dependencies

Also related: https://github.com/input-output-hk/cardano-wallet/pull/2204

Commit:
Add stack2cabal section in stack.yaml

This will make it simpler to regenerate the .cabal files
with cabal-install < 3.4.

Commit:
Regenerate cabal.project{,.freeze}

This was done by:
  stack2cabal -p now

KtorZ commented on stack.yaml:
```diff
@@ -79,3 +79,185 @@ nix:
+
+# !!! This can be dropped when cabal-3.4.0.0 is released !!!
+#
+#+BEGIN_STACK2CABAL
```
I still don't think that this is the right place for this. But let's see.
